Active Record Callbacks > 10. Transaction Callbacks (#380)
campuscode/rails-guides-pt-BR: pt-BR/active_record_callbacks.md

You can declare as many callbacks as you want inside your callback classes.

> Callbacks de Transação

Reviewer: ```suggestion
*Callbacks* de Transação
```

Acho que podemos manter o itálico no título também, o que acha?

Author: Sim, concordo :)

---------------------

> Existem dois *callbacks* adicionais que são gatilhados quando se completa uma transação de banco de dados: `after_commit` e `after_rollback`. Estes *callbacks* são muito parecidos com o *callback* `after_save`, exceto que eles não são executados até que as mudanças no banco de dados sejam confirmadas ou desfeitas. Eles são mais úteis quando seus modelos de registros ativos precisam de interagir com sistemas externos que não fazem parte da transação do banco de dados.

Reviewer 2: ```suggestion
Existem dois *callbacks* adicionais que são disparados quando se completa uma transação de banco de dados: `after_commit` e `after_rollback`. Estes *callbacks* são muito parecidos com o *callback* `after_save`, exceto que eles não são executados até que as mudanças no banco de dados sejam confirmadas ou desfeitas. Eles são mais úteis quando seus *active record models* precisam de interagir com sistemas externos que não fazem parte da transação do banco de dados.
```

Fiquei em dúvida aqui em traduzir *active record models*. O que vocês acham @douglas-oliveira-talkdesk @campuscode/guia-rails-team ?

Reviewer: Acho que podemos manter em inglês por causa do padrão, talvez?

Author: Nice, mantemos em ingles 😄

> Considere, por exemplo, o exemplo anterior onde o modelo `PictureFile` precisa de apagar um arquivo depois que um registro correspondente é destruído. Se algo lançar uma exceção depois que o *callback* `after_destroy` for chamado e a transação for desfeita, o arquivo terá sido deletado e o modelo será deixado em um estado inconsistente. Por exemplo, suponha que `picture_file_2` no codigo abaixo não é valido e o método `save!` lança um erro.

Reviewer 2: ```suggestion
Considere, por exemplo, o exemplo anterior onde o *model* `PictureFile` precisa de apagar um arquivo depois que um registro correspondente é destruído. Se algo lançar uma exceção depois que o *callback* `after_destroy` for chamado e a transação for desfeita, o arquivo terá sido deletado e o *model* será deixado em um estado inconsistente. Por exemplo, suponha que `picture_file_2` no código abaixo não é valido e o método `save!` lança um erro.
```

end
```

NOTE: A opção `:on` especifica quando um *callback* vai ser disparado. Se você
não fornecer a opção `:on` o *callback* será disparado para cada ação.

Já que usar o *callback* `after_commit` para criar, atualizar ou deletar é

> comum, exitem *alias* para as operações:

Reviewer 2: ```suggestion
comum, existem *aliases* para as operações:
```

end
```

> WARNING. Quando uma transação é completada, os *callbacks* `after_commit` ou `after_rollback` são chamados para todos os modelos criados, atualizados ou destruidos em uma transação. No entanto, se uma exceção é lançada em um desses *callbacks*, a exceção vai borbulhar e quaisquer métodos restantes de `after_commit` ou `after_rollback` _não_ serão executados. Assim sendo, se o código do seu *callback* pode lançar uma exceção, você precisará a recuperar e tratá-la dentro do *callback* para que outros callbacks possam ser executados.

Reviewer 2: ```suggestion
WARNING. Quando uma transação é completada, os *callbacks* `after_commit` ou `after_rollback` são chamados para todos os *models* criados, atualizados ou destruidos em uma transação. No entanto, se uma exceção é lançada em um desses *callbacks*, a exceção vai borbulhar e quaisquer métodos restantes de `after_commit` ou `after_rollback` _não_ serão executados. Assim sendo, se o código do seu *callback* pode lançar uma exceção, você precisará  recuperá-lo e tratá-lo dentro do *callback* para que outros *callbacks* possam ser executados.
```

Author: Hm, acredito que o "recuperar" e "tratar" faz referência à "exceção", certo? Nesse sentido, prefiro manter no feminino. O que você acha?

Vou aplicar a alteração no "modelos", para já :)

Reviewer 2: > Hm, acredito que o "recuperar" e "tratar" faz referência à "exceção", certo? Nesse sentido, prefiro manter no feminino. O que você acha?
> 
> Vou aplicar a alteração no "modelos", para já :)

Tem razão, faz sentido :smile:

> WARNING. Quando uma transação é completada, os *callbacks* `after_commit` ou `after_rollback` são chamados para todos os *models* criados, atualizados ou destruidos em uma transação. No entanto, se uma exceção é lançada em um desses *callbacks*, a exceção vai borbulhar e quaisquer métodos restantes de `after_commit` ou `after_rollback` _não_ serão executados. Assim sendo, se o código do seu *callback* pode lançar uma exceção, você precisará recuperá-la e tratá-la dentro do *callback* para que outros callbacks possam ser executados.

Reviewer: ```suggestion
WARNING. Quando uma transação é completada, os *callbacks* `after_commit` ou `after_rollback` são chamados para todos os *models* criados, atualizados ou destruidos em uma transação. No entanto, se uma exceção é lançada em um desses *callbacks*, a exceção vai interromper a execução e quaisquer métodos restantes de `after_commit` ou `after_rollback` _não_ serão executados. Assim sendo, se o código do seu *callback* pode lançar uma exceção, você precisará recuperá-la e tratá-la dentro do *callback* para que outros callbacks possam ser executados.
```

Acho que o "borbulhar" em português não fez tanto sentido o que acha de trocarmos ? Não sei o que seria melhor mas fiz uma sugestão

Author: "Estourar" talvez?

Reviewer 2: > "Estourar" talvez? @HenriqueMorato

É que ja tem o conceito de "lançar" anteriormente na frase ... `se uma exceção é lançada em um ...`
Acho que `interromper a execução` pode ser o que faça mais sentido mesmo.

WARNING. O código executado dentro dos *callbacks* de `after_commit` ou `after_rollback` não está incluido em uma transação.

> WARNING. Usando ambos `after_create_commit` e `after_update_commit` no mesmo modelo irá somente permitir que o último *callback* definido seja efetuado, e irá sobrepor todos os outros.

Reviewer 2: ```suggestion
WARNING. Usando ambos `after_create_commit` e `after_update_commit` no mesmo *model* permitirá somente que o último *callback* definido seja efetuado, sobrepondo todos os outros.
```
